threading/smartptrs.nim

```diff
@@ -87,42 +87,40 @@ proc `$`*[T](p: UniquePtr[T]): string {.inline.} =
 type
   SharedPtr*[T] = object
     ## Shared ownership reference counting pointer.
-    val: ptr tuple[value: T, counter: Atomic[int]]
+    container: ptr tuple[value: T, counter: Atomic[int]]
+
+proc decr[T](p: SharedPtr[T]) {.inline.} =
+  if p.container != nil:
+    # this `fetchSub` returns current val then subs
+    # so count == 1 means we're the last
+    if p.container.counter.fetchSub(1, Release) == 1:
```

Comment:
Returning the current value and then decrementing sounds like an unnatural CPU instruction. Are you sure this generates good code?

Reply:
Most of the atomics I've seen return the value after modification.

I haven't looked at the generated code but figure the C intrinsics are correct. Looks like ARM64 does some sort of register locking and the operations inside are flexible: https://stackoverflow.com/a/35218319 Maybe there's performance implications? I sorta thought they had specific atomic instructions, but sounds like x86_64 does something similar. Though this guy seems a bit different: https://devblogs.microsoft.com/oldnewthing/20220811-00/?p=106963 

The `std/atomics` version does the decrement and returns the result. It took me a bit to figure out the difference. Both work and pass tests.

Reply:
Looks like Rust does it similarly in their `Arc`: https://github.com/rust-lang/rust/blob/a09c1f85f1bc5be9a85e1395e7482caff91260cc/library/alloc/src/sync.rs#L2372

Their `fetch_sub` also does "Subtracts from the current value, returning the previous value": https://github.com/rust-lang/rust/blob/a09c1f85f1bc5be9a85e1395e7482caff91260cc/library/core/src/sync/atomic.rs#L2550C1-L2551C1 .

Reply:
Ah, I think I get it a bit. The `release` mode makes the atomic load strictly ordered, so returning that value (the original) makes sense whereas the atomic store isn't strictly ordered with `release`. 

I think doing `sub_fetch` would need to have `acquire` mode.